### PR TITLE
[group-ib-942] Polygon Integration

### PR DESCRIPTION
**Pull request by gib-tds**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Group-IB TDS Polygon Integration

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 4.5.0
- [x] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

## Demisto Partner?
- [x] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

**Comment by ShahafBenYakir** [comment]
### Integration 
1. I could not make the `export-pcap` or `export-video` to work. Any suggestions?
2. Do you support url reputation in your API? If so let's add it too.
3. In many places in the code I commented about the expressions like this: `value = item["key"]` after several occurrences I stoped. Though we could be very certain that a key exist in the response, we highly recommend no to work with such key extraction, rather chagne it to `value = item.get("key")` and then check that the value exists. We had countless bugs opening on such errors since sometimes there are unpredictable customer environment.
### Unit Tests
1. Let's divide the data to file to responses

**Comment by ShahafBenYakir** [comment]
**First of all amazing work! Very impressive flows you have here. Few things unrelated to code:**

#### Integration 
1. I could not make the `export-pcap` or `export-video` to work. Any suggestions?
2. Any suggestion on what to run in order to get an analysis that includes a network/monitor indicator data? I would want to see that `Process` and `RegistryKey` classes in action.
3. can you give a short explanation of what are these types (Package/Monitor/Netowork)?
4. Do you support URL reputation in your API? If so let's add a command fo it too.
5. In many places in the code I commented about expressions like this: `value = item["key"]` after several occurrences I stopped as it appears a lot. 
Though we could be very certain that a key exists in some item/dict, we highly recommend no to work with such key extraction operation as there are cases we haven't thought about and always exists especially in customer environments that tend to be slightly different than ours. Rather we encourage the use of the `get` operation, like this: `value = item.get("key")` then you can check that the value does exist. We had countless bugs opening on such errors so this is really a lesson we have learned the hard way.

#### Unit Tests
1. Let's divide the `mocked_data` file into several files, one for each relevant server response.
2. Let's also make sure that each such file includes **only the needed data to the test it's been used for**. Currently, the data file is 15,000 lines, I don't see a reason why it should get bigger than a hundred lines (no need to cover cases of big lists in any of the functions - if we want to test some behavior on a list, 2-3 items should be enough).
3. Let's separate the mock data from the mock response. 

#### Detonation Playbooks 
1. Both playbooks fail on benign samples, need to fix that.
2. We need to add another condition to the conditional task that checks for the dump existence. When there is no dump there is a type error. Consider this addition: 
Playbooks 2020-06-07 19-53-47

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+        'Status': 'In Progress',
+    }
+    results = CommandResults(
+        readable_output=f"Url uploaded successfully. Analysis ID: {res}",
+        outputs_prefix='Polygon.Analysis',
+        outputs_key_field='ID',
+        outputs=outputs
+    )
+    return_results(results)
+    return results
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:504]: If you wish to test the command function you can use the structure used in the `HelloWorld_test`. Which is moving the `return_results` call to the main function, then in your test you can assert the CommandResults returned from this command.

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+        resp = requests.request(
+            method,
+            url,
+            headers=self.headers,
+            verify=self.verify,
+            params=params,
+            data=data,
+            files=files,
+            proxies=self.proxies
+        )
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+        PARSE AND VALIDATE INTEGRATION PARAMS
+    """
+    params = demisto.params()
+    base_url = params.get('server')
+    api_key = params.get('api_key')
+    verify_certificate = not params.get('insecure', False)
+    report_language = LANGUAGE_TO_CODE[params.get('report_language')]
+
+    # Remove proxy if not set to true in params
+    proxies = handle_proxy()
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+
+    def get_attach(self, id=None):
+        url = ATTACH.format(id) if id else ATTACHES
+        resp = self._http_request('get', url)["data"]["results"]
+        if id:
+            try:
+                resp = resp[0]
+            except Exception:
+                raise ValueError(f"File with ID={id} does not exist")
+        return resp
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done in all such places

File: Packs/Polygon/Integrations/Polygon/Polygon.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:475]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+        'Status': 'In Progress'
+    }
+    results = CommandResults(
+        readable_output=f"File uploaded successfully. Analysis ID: {res}",
+        outputs_prefix='Polygon.Analysis',
+        outputs_key_field='ID',
+        outputs=outputs
+    )
+    return_results(results)
+    return results
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:499]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.yml [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+          description: The information about IP indicators
+        - contextPath: Polygon.Analysis.Domain
+          description: The information about Domain indicators
+        - contextPath: Polygon.Analysis.RegistryKey
+          description: The information about registry keys which were modified during
+            the analysis
+        - contextPath: Polygon.Analysis.Process
+          description: The information about processes started during the analysis
+      description: Get TDS Polygon analysis info
+    - name: polygon-report-export
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.yml]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+            proxies=proxies,
+            language=report_language
+        )
+        commands = {
+            'polygon-upload-file': upload_file_command,
+            'polygon-upload-url': upload_url_command,
+            'polygon-analysis-info': analysis_info_command,
+            'polygon-report-export': export_report_command,
+            'polygon-pcap-export': export_pcap_command,
+            'polygon-video-export': export_video_command,
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon_description.md [hunk 8]
```diff
@@ -0,0 +1,6 @@
+ ## Group-IB TDS Polygon
+- This section explains how to configure the instance of Polygon in Cortex XSOAR.
+
+1. Open TDS Huntbox web interface.
+2. Navigate to Profile and click Generate auth token.
+3. Copy this token. This is your API Key.
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon_description.md]: Done

File: Packs/Polygon/Playbooks/playbook-Detonate_URL_-_Group-IB_TDS_Polygon_README.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+| Polygon.Analysis.DumpExists | boolean | Network activity dump exists | 
+| Polygon.Analysis.File | unknown | The information about files in analysis | 
+| Polygon.Analysis.URL | unknown | The information about URL indicators | 
+| Polygon.Analysis.IP | unknown | The information about IP indicators | 
+| Polygon.Analysis.Domain | unknown | The information about Domain indicators | 
+| Polygon.Analysis.RegistryKey | unknown | The information about registry keys which were modified during the analysis | 
+| Polygon.Analysis.Process | unknown | The information about processes started during the analysis | 
+
+## Playbook Image
+![Polygon Detonate URL](https://github.com/demisto/content/raw/Polygon-Initial/docs/images/playbooks/Detonate_URL_-_Group-IB_TDS_Polygon.png)
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Playbooks/playbook-Detonate_URL_-_Group-IB_TDS_Polygon_README.md]: Done

File: Packs/Polygon/Playbooks/playbook-Detonate_File_-_Group-IB_TDS_Polygon_README.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+| Polygon.Analysis.DumpExists | boolean | Network activity dump exists | 
+| Polygon.Analysis.File | unknown | The information about files in analysis | 
+| Polygon.Analysis.URL | unknown | The information about URL indicators | 
+| Polygon.Analysis.IP | unknown | The information about IP indicators | 
+| Polygon.Analysis.Domain | unknown | The information about Domain indicators | 
+| Polygon.Analysis.RegistryKey | unknown | The information about registry keys which were modified during the analysis | 
+| Polygon.Analysis.Process | unknown | The information about processes started during the analysis | 
+
+## Playbook Image
+![Polygon Detonate File](https://github.com/demisto/content/raw/Polygon-Initial/docs/images/playbooks/Detonate_File_-_Group-IB_TDS_Polygon.png)
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Playbooks/playbook-Detonate_File_-_Group-IB_TDS_Polygon_README.md]: Done

**Comment by ShahafBenYakir** [comment]
In the test playbook in the `file` 
Change this:
Playbooks 2020-06-14 17-49-09

To this:
Playbooks 2020-06-14 17-46-18

**Comment by gib-tds** [comment]
> In the test playbook in the `file`
> Change this:
> Playbooks 2020-06-14 17-49-09
> 
> To this:
> Playbooks 2020-06-14 17-46-18

Done

File: Packs/Polygon/Playbooks/playbook-Detonate_File_-_Group-IB_TDS_Polygon.yml [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+      - "1"
+    scriptarguments:
+      value:
+        complex:
+          root: modules
+          filters:
+          - - operator: isEqualString
+              left:
+                value:
+                  simple: modules.name
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Playbooks/playbook-Detonate_File_-_Group-IB_TDS_Polygon.yml]: Done

File: Packs/Polygon/Playbooks/playbook-Detonate_URL_-_Group-IB_TDS_Polygon.yml [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+      - "18"
+    scriptarguments:
+      value:
+        complex:
+          root: modules
+          filters:
+          - - operator: isEqualString
+              left:
+                value:
+                  simple: modules.name
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Playbooks/playbook-Detonate_URL_-_Group-IB_TDS_Polygon.yml]: Done

File: Packs/Polygon/TestPlaybooks/playbook-Polygon-Test.yml [hunk 13]
```diff
@@ -0,0 +1,939 @@
...
+    ignoreworker: false
+  "12":
+    id: "12"
+    taskid: 9e50da12-6ecb-4787-8755-df749edc01bf
+    type: regular
+    task:
+      id: 9e50da12-6ecb-4787-8755-df749edc01bf
+      version: -1
+      name: polygon-video-export
+      script: '|||polygon-video-export'
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/TestPlaybooks/playbook-Polygon-Test.yml]: Done

File: Packs/Polygon/TestPlaybooks/playbook-Polygon-Test.yml [hunk 14]
```diff
@@ -0,0 +1,939 @@
...
+    ignoreworker: false
+  "10":
+    id: "10"
+    taskid: 41edae72-2804-4f14-807f-9aa8ec995183
+    type: regular
+    task:
+      id: 41edae72-2804-4f14-807f-9aa8ec995183
+      version: -1
+      name: polygon-pcap-export
+      script: '|||polygon-pcap-export'
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/TestPlaybooks/playbook-Polygon-Test.yml]: Done

File: Packs/Polygon/TestPlaybooks/playbook-Polygon-Test.yml [hunk 15]
```diff
@@ -0,0 +1,939 @@
...
+    ignoreworker: false
+  "8":
+    id: "8"
+    taskid: 034ee9f3-88ba-4905-8b3a-a3c5472b92c4
+    type: regular
+    task:
+      id: 034ee9f3-88ba-4905-8b3a-a3c5472b92c4
+      version: -1
+      name: polygon-report-export
+      script: '|||polygon-report-export'
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/TestPlaybooks/playbook-Polygon-Test.yml]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 16]
```diff
@@ -0,0 +1,622 @@
...
+        if "analgin_result" in file_info:
+            if "commit" in file_info["analgin_result"]:
+                if "reports" in file_info["analgin_result"]:
+                    if len(file_info["analgin_result"]["reports"]):
+                        if "id" in file_info["analgin_result"]["reports"][0]:
+                            report = True
+        return report
+
+    def _get_fids(self, resp):
+        fids = resp["data"].get("ids", [])
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 17]
```diff
@@ -0,0 +1,622 @@
...
+        )
+        return self._resp_handle(resp, decode)
+
+    def _check_report_available(self, file_info):
+        report = False
+        if "analgin_result" in file_info:
+            if "commit" in file_info["analgin_result"]:
+                if "reports" in file_info["analgin_result"]:
+                    if len(file_info["analgin_result"]["reports"]):
+                        if "id" in file_info["analgin_result"]["reports"][0]:
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 18]
```diff
@@ -0,0 +1,622 @@
...
+        raise ValueError("No reports found")
+
+    def export_report(self, tds_analysis_id):
+        file = self.get_attach(tds_analysis_id)
+        if self._check_report_available(file):
+            return self._http_request(
+                method='get',
+                url_suffix=REPORT_EXPORT.format(tds_analysis_id,
+                                                file["analgin_result"]["commit"],
+                                                file["analgin_result"]["reports"][0]["id"]),
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:173]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 19]
```diff
@@ -0,0 +1,622 @@
...
+        raise ValueError("No reports found")
+
+    def export_pcap(self, tds_analysis_id):
+        file = self.get_attach(tds_analysis_id)
+        if self._check_report_available(file):
+            return self._http_request(
+                method='get',
+                url_suffix=PCAP_EXPORT.format(tds_analysis_id,
+                                              file["analgin_result"]["commit"],
+                                              file["analgin_result"]["reports"][0]["id"]),
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:185]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 20]
```diff
@@ -0,0 +1,622 @@
...
+
+    def export_video(self, tds_analysis_id):
+        file = self.get_attach(tds_analysis_id)
+        if self._check_report_available(file):
+            try:
+                return self._http_request(
+                    method='get',
+                    url_suffix=VIDEO_EXPORT.format(tds_analysis_id,
+                                                   file["analgin_result"]["commit"],
+                                                   file["analgin_result"]["reports"][0]["id"]),
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 21]
```diff
@@ -0,0 +1,622 @@
...
+                )
+            except ValueError:
+                return None
+        raise ValueError("No dump files found")
+
+    def get_hash_reputation(self, hash_type, hash_value):
+        return self._http_request(
+            method='GET',
+            url_suffix=HASH_REPUTATION.format(hash_type, hash_value)
+        )["data"]
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 22]
```diff
@@ -0,0 +1,622 @@
...
+    return ids
+
+
+def get_network_indicators(report):
+    ids: List[Common.Indicator] = []
+    network = report['network']
+    for dns in network['dns']:
+        domain = Common.Domain(
+            domain=dns['request'],
+            dns=", ".join(list(map(lambda x: x['data'], dns['answers']))),
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 23]
```diff
@@ -0,0 +1,622 @@
...
+            analysis_info = {
+                hash_type.upper(): file,
+                'Found': res.get('found'),
+                'Verdict': res.get('verdict'),
+                'Score': res.get('score'),
+                'Malware-families': res.get('malware_families')
+            }
+            score = Common.DBotScore.NONE
+            malicious = None
+            if res["found"]:
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 24]
```diff
@@ -0,0 +1,622 @@
...
+                if res["verdict"]:
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:173]: We know that fields exist as it was checked earlier

File: Packs/Polygon/Integrations/Polygon/Polygon.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:185]: We know that fields exist as it was checked earlier

File: Packs/Polygon/Integrations/Polygon/Polygon.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: We know that fields exist as it was checked earlier

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 25]
```diff
@@ -0,0 +1,624 @@
...
+        results = CommandResults(
+            readable_output=human_readable,
+            outputs_prefix="Polygon.Analysis",
+            outputs_key_field="ID",
+            outputs=analysis_info,
+            indicators=indicators
+        )
+        all_results.append(results)
+        return_results(results)
+    return all_results
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:424]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 26]
```diff
@@ -0,0 +1,624 @@
...
+
+
+def export_report_command(client, args):
+    tds_analysis_id = drop_prefix(args.get('tds_analysis_id'))
+    report = client.export_report(tds_analysis_id)
+    demisto.results(fileResult(
+        filename='report.tar',
+        data=report,
+        file_type=EntryType.ENTRY_INFO_FILE
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:433]: No need for `file_type` or all 4 lines?

File: Packs/Polygon/Integrations/Polygon/Polygon.yml [hunk 27]
```diff
@@ -0,0 +1,356 @@
...
+          type: string
+        - contextPath: InfoFile.Size
+          description: The dump file size
+          type: number
+        - contextPath: InfoFile.Type
+          description: The dump file type
+          type: string
+        - contextPath: InfoFile.Info
+          description: The dump file info
+      description: Network activity dump export
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.yml]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.yml [hunk 28]
```diff
@@ -0,0 +1,356 @@
...
+        - contextPath: InfoFile.Size
+          description: The video file size
+          type: number
+        - contextPath: InfoFile.Type
+          description: The video file type
+          type: string
+        - contextPath: InfoFile.Info
+          description: The video file info
+          type: string
+      description: Screen activity video export
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.yml]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 29]
```diff
@@ -0,0 +1,624 @@
...
+            language=report_language
+        )
+        commands = {
+            'polygon-upload-file': upload_file_command,
+            'polygon-upload-url': upload_url_command,
+            'polygon-analysis-info': analysis_info_command,
+            'polygon-report-export': export_report_command,
+            'polygon-pcap-export': export_pcap_command,
+            'polygon-video-export': export_video_command,
+            'file': file_command
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py:578]: Unfortunately there is no support for URL reputation

File: Packs/Polygon/Integrations/Polygon/Polygon.yml [hunk 30]
```diff
@@ -0,0 +1,356 @@
...
+        - contextPath: InfoFile.Size
+          description: The dump file size
+          type: number
+        - contextPath: InfoFile.Type
+          description: The dump file type
+          type: string
+        - contextPath: InfoFile.Info
+          description: The dump file info
+      description: Network activity dump export
+    - name: polygon-video-export
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.yml]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.yml [hunk 31]
```diff
@@ -0,0 +1,356 @@
...
+          description: The report size
+          type: number
+        - contextPath: InfoFile.Type
+          description: The report file type
+          type: string
+        - contextPath: InfoFile.Info
+          description: The report file info
+          type: string
+      description: Export an archive with TDS Polygon report to War Room
+    - name: polygon-pcap-export
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.yml]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 32]
```diff
@@ -0,0 +1,622 @@
...
+        'ID': res,
+        'URL': url,
+        'Status': 'In Progress',
+    }
+    results = CommandResults(
+        readable_output=f"Url uploaded successfully. Analysis ID: {res}",
+        outputs_prefix='Polygon.Analysis',
+        outputs_key_field='ID',
+        outputs=outputs
+    )
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 33]
```diff
@@ -0,0 +1,622 @@
...
+    url = args.get('url')
+    res = client.upload_url(url)
+    res = f"{URL_TYPE}{res}"
+    outputs = {
+        'ID': res,
+        'URL': url,
+        'Status': 'In Progress',
+    }
+    results = CommandResults(
+        readable_output=f"Url uploaded successfully. Analysis ID: {res}",
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 34]
```diff
@@ -0,0 +1,622 @@
...
+    res = client.upload_file(file_name, file_path, password)
+    res = f"{FILE_TYPE}{res}"
+    outputs = {
+        'ID': res,
+        'EntryID': file_id,
+        'FileName': file_obj['name'],
+        'Status': 'In Progress'
+    }
+    results = CommandResults(
+        readable_output=f"File uploaded successfully. Analysis ID: {res}",
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 35]
```diff
@@ -0,0 +1,624 @@
...
+        return resp
+
+    def get_url(self, tds_analysis_id):
+        file = self.get_attach(tds_analysis_id)
+        if self._check_report_available(file):
+            return self._http_request(
+                method='get',
+                url_suffix=file["file_url"][1:],
+                decode=False
+            )
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 36]
```diff
@@ -0,0 +1,624 @@
...
+    tds_analysis_id_array = argToList(args.get('tds_analysis_id'))
+    all_results = []
+    for tds_analysis_id in tds_analysis_id_array:
+        analysis_type = tds_analysis_id[0]
+        res = client.get_analysis_info(drop_prefix(tds_analysis_id))
+        analysis_info = serialize_analysis_info(res['file'], analysis_type, report='report' in res)
+        indicators = []
+        if 'report' in res:
+            if analysis_type == URL_TYPE:
+                res['report']['target']['url'] = client.get_url(drop_prefix(tds_analysis_id)).decode()
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 37]
```diff
@@ -0,0 +1,624 @@
...
+        'ID': res,
+        'EntryID': file_id,
+        'FileName': file_obj['name'],
+        'Status': 'In Progress'
+    }
+    results = CommandResults(
+        readable_output=f"File uploaded successfully. Analysis ID: {res}",
+        outputs_prefix='Polygon.Analysis',
+        outputs_key_field='ID',
+        outputs=outputs
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 38]
```diff
@@ -0,0 +1,624 @@
...
+                res['report']['target']['url'] = client.get_url(drop_prefix(tds_analysis_id)).decode()
+            analysis_info.update(serialize_report_info(res['report'], analysis_type))
+            indicators = get_report_indicators(res["report"], analysis_type)
+        human_readable = get_human_readable_analysis_info(analysis_info)
+        results = CommandResults(
+            readable_output=human_readable,
+            outputs_prefix="Polygon.Analysis",
+            outputs_key_field="ID",
+            outputs=analysis_info,
+            indicators=indicators
```
[the remark this answers is not in the document]
Comment (gib-tds) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: Done

**Comment by ShahafBenYakir** [comment]
@gib-tds 
For `polygon-export-report`, `polygon-export-video` and `polygon-export-pcap` all context outputs should be under `File` and not `InfoFile`. For example `File.Name`, `File.EntryID` and so on...
This also needs to be fixed in the test playbook that currently fails because of that.

File: Packs/Polygon/Integrations/Polygon/Polygon.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Packs/Polygon/Integrations/Polygon/Polygon.py]: file_type

File: Packs/Polygon/Integrations/Polygon/Polygon.py [hunk 39]
```diff
@@ -0,0 +1,598 @@
...
+        'Status': 'In Progress',
+    }
+    results = CommandResults(
+        readable_output=tableToMarkdown("Url uploaded successfully", outputs),
+        outputs_prefix='Polygon.Analysis',
+        outputs_key_field='ID',
+        outputs=outputs,
+        raw_response=res
+    )
+    return results
```
Comment (ShahafBenYakir) [inline Packs/Polygon/Integrations/Polygon/Polygon.py:473]: Did we choose to leave this on purpose?

**Comment by gib-tds** [comment]
> @gib-tds
> For `polygon-export-report`, `polygon-export-video` and `polygon-export-pcap` all context outputs should be under `File` and not `InfoFile`. For example `File.Name`, `File.EntryID` and so on...
> This also needs to be fixed in the test playbook that currently fails because of that.
> This is the only thing left in my opinion so after you are done just run the test PB to make sure it works.

Done it! Test Playbook works

**Pull request merged**